kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+/** Detail class definition.
+*/
+template <class TPointerDataType, class TValueDataType>
+class GlobalPointerMapCommunicator
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using GlobalPointerType = GlobalPointer<TPointerDataType>;
+
+    using GlobalPointerVectorType = std::vector<GlobalPointerType>;
+
+    using GlobalPointerMapType = std::unordered_map<int, GlobalPointerVectorType>;
+
+    using DataVectorType = std::vector<TValueDataType>;
+
+    using DataVectorMapType = std::unordered_map<int, DataVectorType>;
+
+    template<class TApplyFunctor>
+    using ProxyType = ApplyProxy<TPointerDataType, TValueDataType, TApplyFunctor>;
+
+    /// Pointer definition of GlobalPointerMapCommunicator
+    KRATOS_CLASS_POINTER_DEFINITION(GlobalPointerMapCommunicator);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * @brief Construct a new Global Pointer Map Communicator object
+     *
+     * This constructor should only be called in non-parallel regions
+     *
+     * @param rDataCommunicator         Data communicator
+     */
+    GlobalPointerMapCommunicator(
+        const DataCommunicator& rDataCommunicator)
+        : mrDataCommunicator(rDataCommunicator),
+          mCurrentRank(rDataCommunicator.Rank())
+    {
+    }
+
+    /// Destructor.
+    virtual ~GlobalPointerMapCommunicator() = default;
+
+    /// Assignment constructor
+    GlobalPointerMapCommunicator& operator=(GlobalPointerMapCommunicator const& rOther)  = delete;
+
+    /// Copy constructor.
+    GlobalPointerMapCommunicator(GlobalPointerMapCommunicator const& rOther) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Sends and applies remote Gp values
+     *
+     * This method is used to do communication between processes to apply non_local gp
+     * values in each process in GP's owning process
+     *
+     * Ghost mesh synchronization needs to be done afterwards.
+     *
+     * @tparam TApplyFunctor
+     * @param rApplyProxy           The proxy which holds the user specified methods
+     */
+    template <class TApplyFunctor>
+    void SendAndApplyRemotely(ProxyType<TApplyFunctor>& rApplyProxy)
+    {
+        KRATOS_TRY
+
+        if (IsDistributed()) {
+
+            KRATOS_DEBUG_ERROR_IF(OpenMPUtils::IsInParallel() != 0)
+                << "Calling SendAndApplyRemotely in a parallel region is not "
+                   "allowed.\n";
+
+            // get the final map for communications
+            // constructing an rank based GlobalPointersUnorderedMap will make unique keys list in GlobalPointersUnorderedMap
+            // which will result in lower communication for keys. These unique keys can be easily OMP parallelized
+            std::unordered_map<int, GlobalPointersUnorderedMap<TPointerDataType, std::vector<TValueDataType>>> non_local_map;
+
+            for (IndexType i = 0; i < rApplyProxy.mNonLocalGlobalPointerMapsVector.size(); ++i) {
```

Comment:
use reserve before push_back if you know the exact size

Reply:
Actually the problem is we don't know the exact size because, when the if this is run in OMP, then `mNonLocalGlobalPointerMapsVector` is of size `NumberOfThreads`, and inside each of its elements, there is a map with keys=rank, and values=gps. So at this point we don't know exactly how many ranks are included in the OMP loops. :/